### DISCUSSION TITLE
HTML Entities Node

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#17] [Ideas]
I've noticed that PML doesn't support inline SGML char entities, which makes sense in terms of a clean spec. It might be worth adding their support via a dedicated inline node, along the line of those that handle special characters insertions:

- `[&#<dec.val>]`
- `[&#x<hex.val>]`
- `[&<name>]`

E.g.

|    PML    |   HTML   |  char  |    type   |
|-----------|----------|--------|-----------|
| `[&#60]`  | `&#60;`  | &#60;  | dec value |
| `[&#x3c]` | `&#x3c;` | &#x3c; | hex value |
| `[&lt]`   | `&lt;`   | &lt;   | named     |

Surely, using a tag makes each entity one char longer, but that doesn't seem a big deal since editors will handle auto-bracket-matching, and the tag notation looks cleaner.

Furthermore, users can exploit parameters inside the `[&` tag, if the need be.

Modern editors can also render the corresponding char on mouse hovering, or simply inject it after/inside the tag, for previewing purposes.

-------------------------------------------------------------------------------

# References

- [Wikipedia » SGML entity](https://en.wikipedia.org/wiki/SGML_entity)
- [Wikipedia » List of XML and HTML character entity references](https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references)

**Comment by pxml-lang** [comment]
> PML doesn't support inline SGML char entities

The current solution (workaround) is to use a [verbatim](https://www.pml-lang.dev/docs/reference_manual/index.html#node_verbatim) tag. For example, the following PML code:
```
[verbatim &#60;] [verbatim &#x3c] [verbatim &lt;]
```
... will be rendered as:
```
< < <
```
This works as long as PML's output is HTML.

When the new pXML parser is used in PML, it will also be possible to use the `\uhhhh` syntax to define a Unicode hexadecimal value. The following PML code:
```
\u003c
```
... will be rendered as:
```
<
```
This will work in all output formats (not only HTML), as the escape sequence is transformed into a Unicode character by the parser.

PML's counterpart for SGML entities are constants. The following SGML entity declarations (borrowed from [Wikipedia](https://en.wikipedia.org/wiki/SGML_entity)):
```
<!ENTITY greeting1 "Hello world">
<!ENTITY greeting2 SYSTEM "file:///hello.txt">
<!ENTITY % greeting3 "¡Hola!">
<!ENTITY greeting4 "%greeting3; means Hello!">
```
... will be written like this in PML:
```
[const greeting1 = Hello world]
[const greeting2 read file hello.txt]    // 'read url' will also be supported
[const greeting3 = ¡Hola!]
[const greeting4 = [= greeting3] means Hello!]
```
(Note that the final syntax might be slightly different.)
The constants can then be used in the text like this:
```
[= greeting1]
[= greeting2]
[= greeting3]
[= greeting4]
```
... which will be rendered as:
```
Hello world
<content of file hello.txt>
¡Hola!
¡Hola! means Hello!
```
A user could also define:
```
[const lt = \u003c]
```
... and then use it like this:
```
[= lt]
```
to 'simulate' HTML entities that will work in all output formats (text, PDF, Open Document, etc.).

With all this in place I wonder if we still need a dedicated syntax for SGML char entities. What do you think?

**Reply by tajmone** [reply]
The `\uhhhh` notation seems a very good solution indeed; also, it's easier to scope since it's just another escape sequence.

So eventually parameters will switch from the `<<`…`>>` notation to `[=`…`]`? That's good, I like it better.
